odoo/models.py

```diff
@@ -2650,9 +2793,125 @@ def _inherits_join_calc(self, alias, fname, query):
+                    query._raw_joins[expr] = []
+
+            return property_alias
+
+        elif property_type == 'selection':
+            options = definition.get('selection') or []
+            options = [option[0] for option in options]
+
+            # check the existence of the option
+            expr = f'''
+                LEFT JOIN (SELECT unnest(%s::text[]) {property_alias}) {property_alias}
+                       ON {property_sql}->>0 = {property_alias}
+            '''
+            query._raw_joins[expr] = [options]
+
+            return property_alias
+
+        elif property_type == 'many2one':
+            comodel = self.env.get(definition.get('comodel'))
+            if comodel is None or comodel._transient or comodel._abstract:
+                # all value are false, because the model does not exist anymore
+                # (or is a transient model e.g.)
+                return 'FALSE'
+
+            return f'''
+                CASE
+                    WHEN jsonb_typeof({property_sql}) = 'number'
+                     AND ({property_sql})::int IN (SELECT id FROM {comodel._table})
+                    THEN {property_sql}
+                    ELSE NULL
+                 END
+            '''
+
+        elif property_type == 'date':
+            return f'''
+                CASE
+                    WHEN jsonb_typeof({property_sql}) = 'string'
+                    THEN ({property_sql}->>0)::DATE
+                    ELSE NULL
+                 END
+            '''
+
+        elif property_type == 'datetime':
+            return f'''
+                CASE
+                    WHEN jsonb_typeof({property_sql}) = 'string'
+                    THEN to_timestamp({property_sql}->>0, 'YYYY-MM-DD HH24:MI:SS')
+                    ELSE NULL
+                 END
+            '''
+
+        # if the key is not present in the dict, fallback to false instead of none
+        return f"COALESCE({property_sql}, 'false')"
+
+    @api.model
+    def get_property_definition(self, full_name):
+        """Return the definition of the given property.
+
+        :param full_name: Name of the field / property
+            (e.g. "property.integer")
+        """
+        self.check_access_rights("read")
+        field_name, property_name = full_name.split(".")
+        check_property_field_value_name(property_name)
+        if field_name not in self._fields:
+            raise ValueError(f"Wrong field name {field_name!r}.")
+
+        field = self._fields[field_name]
+        target_model = self.env[self._fields[field.definition_record].comodel_name]
+        query = """
+            SELECT definition
+              FROM {table},
+                   jsonb_array_elements({field}) definition
+             WHERE {field} IS NOT NULL
+               AND definition->>'name' = %s
+             LIMIT 1
+        """.format(table=target_model._table, field=field.definition_record_field)
+        self.env.cr.execute(query, [property_name])
```

Comment:
This will have to be written using sql.Identifiers and named parameters I think, like
```suggestion
        query = sql.SQL("""
            SELECT definition
              FROM {table},
                   jsonb_array_elements({field}) definition
             WHERE {field} IS NOT NULL
               AND definition->>'name' = %{property_name}s
             LIMIT 1
        """.format(table=sql.Indentifier(target_model._table), field=sql.Indentifier(field.definition_record_field))
        self.env.cr.execute(query, {'property_name': property_name})
```